### ISSUE TITLE
Background Regions Falsely Highlighted in Heatmaps Despite Strong Image-Level Anomaly Metrics

### ISSUE THREAD

**Issue by YasuoFly** [xiaozhen228/Bayes-PFL#9]
Thanks for your awesome job！

Despite achieving very strong image-level anomaly detection metrics on my custom dataset, the generated heat maps consistently highlight background regions as “anomalous.” I even removed all background content (setting it to pure black), but the heat maps still light up those areas. Note that this is an image-level (no-segmentation-mask) task, so I don’t have pixel-level ground truth to constrain the model.

following is my metrics and heatmap, in order to get the heatmap so i used calcuate_metric_pixel while test and use "best_threshold_cls" as threshold , train on "only classification".

Thank you for any insights!

![Image](https://github.com/user-attachments/assets/18337569-ddf0-468a-881a-60f0b67f17a0)

![Image](https://github.com/user-attachments/assets/d3506a60-c053-4b9f-958c-7cb28b92c741)

**Comment by LonglongaaaGo** [comment]
Hey @xiaozhen228 ，

Thank you so much for your awesome project. I have the same issue. Could you give some suggestions for solving this problem?

Thank you in advance!

**Comment by xiaozhen228** [comment]
We hypothesize that this issue may be attributed to the inherent limitations of the patch-level features in the original pretrained CLIP model, which tend to induce hallucination effects. Even when background pixels are masked out (e.g., set to zero), similar issues may persist due to the global self-attention mechanism of the Transformer architecture.  As our work focuses mainly on the exploration of text prompts, fine-grained visual features were not a primary consideration.  Some previous works have conducted similar explorations, which may provide useful insights for further improvements：
[1] CLIP Surgery for Better Explainability with Enhancement in Open-Vocabulary Tasks
[2] AnomalyCLIP: Object-agnostic Prompt Learning for Zero-shot Anomaly Detection (V-V attention)
[3] Anovl: Adapting vision-language models for unified zero-shot anomaly localization

**Comment by LonglongaaaGo** [comment]
Hey @xiaozhen228 ,

Thanks!
So, if the performance is good but the attention map is located out of the object, is that performance trustworthy?  How do you think about this issue? 

I appreciate your valuable comments and suggestions.
Thank you again!

**Comment by xiaozhen228** [comment]
> Hey [@xiaozhen228](https://github.com/xiaozhen228) ,
> 
> Thanks! So, if the performance is good but the attention map is located out of the object, is that performance trustworthy? How do you think about this issue?
> 
> I appreciate your valuable comments and suggestions. Thank you again!

Hello, in my experiments, such situations rarely occur for most categories, especially for texture-related ones. Moreover, some evaluation metrics (e.g., AUROC and AP) focus more on the relative ranking of anomaly scores rather than their absolute values, so the impact on the evaluation results is limited. Finally, it is worth noting that we adopt a per-anomaly-map min-max normalization strategy for visualization. Therefore, another possible cause of the visualization artifacts might be the noise introduced during this normalization process.

**Comment by YasuoFly** [comment]
@xiaozhen228 
Thanks for your quickly reply!
I tried training on the headCT dataset mentioned in the paper and also tested on headCT. The metrics and visual results are shown below. Is this expected behavior?

<img width="894" height="113" alt="Image" src="https://github.com/user-attachments/assets/d4173fb3-4fe2-402d-9f1a-e82289695750" />

<img width="1054" height="256" alt="Image" src="https://github.com/user-attachments/assets/7aa62893-6ae6-41a7-9660-78e23b1f779f" />

I also tried using the VISA pre-trained model you provided. The metrics are slightly worse than the model trained directly on headCT, but the visual results are significantly better. Could you please help me understand why this is the case?

<img width="890" height="123" alt="Image" src="https://github.com/user-attachments/assets/3c0424db-c1d3-4b4a-af4e-a6f1c30bca9f" />

<img width="1054" height="256" alt="Image" src="https://github.com/user-attachments/assets/7c454615-a0ff-46cb-8325-9afbcbe03a43" />

Thank you for your patience!

**Comment by xiaozhen228** [comment]
**Q1:** I tried training on the headCT dataset mentioned in the paper and also tested on headCT. The metrics and visual results are shown below. Is this expected behavior?
**A1:**  When the training and testing datasets are the same, the classification and segmentation performance is theoretically expected to be better. However, the headCT dataset appears to contain only ground-truth labels for classification, but not for segmentation. Therefore, I suspect that segmentation-related losses (e.g., focal loss and dice loss) were not used during training with the headCT dataset? 


**Q2:** I also tried using the VISA pre-trained model you provided. The metrics are slightly worse than the model trained directly on headCT, but the visual results are significantly better. Could you please help me understand why this is the case?
**A2:** Training on VisA and testing on other datasets (e.g. headCT) follows the zero-shot setting we adopt. First, due to the significant domain gap between different datasets and categories, the performance is expected to drop. During training on VisA, we use both segmentation and classification losses, which contributes to better visualization performance on other datasets.

**Comment by YasuoFly** [comment]
@xiaozhen228 
Thank you very much for your clarification.

Sorry, I forgot to mention earlier. In the paper, the entire headCT dataset was used as a test set to evaluate generalization. For a simple verification on my side, I split the headCT dataset roughly in half into a training set and a test set. I trained on the training set and tested on the test set to obtain the results shown in Q1. That is, the training and test sets are not the same.

Since only classification labels are available, I followed the code and set the ground truth (gt) as all-black images matching the size of the input images. The results shown in Q2 are tested on the test split I created.

My question is: is the model trained in Q1 under this setup reliable or trustworthy?

**Comment by xiaozhen228** [comment]
> [@xiaozhen228](https://github.com/xiaozhen228) Thank you very much for your clarification.
> 
> Sorry, I forgot to mention earlier. In the paper, the entire headCT dataset was used as a test set to evaluate generalization. For a simple verification on my side, I split the headCT dataset roughly in half into a training set and a test set. I trained on the training set and tested on the test set to obtain the results shown in Q1. That is, the training and test sets are not the same.
> 
> Since only classification labels are available, I followed the code and set the ground truth (gt) as all-black images matching the size of the input images. The results shown in Q2 are tested on the test split I created.
> 
> My question is: is the model trained in Q1 under this setup reliable or trustworthy?

1.  I believe that under the setup in Q1, the model is reliable for classification but not for segmentation. This is because assigning all pixel-level ground truth values to 0 (i.e., background) is incorrect for anomaly segmentation. When the ground truth for the entire training set is set to background, the model learns an erroneous segmentation behavior, as reflected in the visualizations shown in Q1. It is important to note that the accuracy of anomaly map visualizations depends on correct segmentation, not just classification.

2.  The setup in Q1 seems to follow a conventional supervised classification setting without considering cross-category generalization (i.e., zero-shot). If the model only focuses on classifying the same categories seen during training, I would suggest removing the segmentation loss. However, this would prevent the generation of meaningful anomaly maps.  In a word, if visualizing anomaly maps is desired, the auxiliary training set must contain both normal and anomalous samples, and the segmentation loss must be optimized with correct pixel-level ground truth.

**Comment by LonglongaaaGo** [comment]
Hey @xiaozhen228 ,

Thank you so much!
So in this case, as @YasuoFly mentioned. To train on the data without pixel-level labels, removing the segmentation loss and DICE loss would be a much better choice, right?
I am surprised your method is doing so well for the classification.

**Comment by xiaozhen228** [comment]
> Hey [@xiaozhen228](https://github.com/xiaozhen228) ,
> 
> Thank you so much! So in this case, as [@YasuoFly](https://github.com/YasuoFly) mentioned. To train on the data without pixel-level labels, removing the segmentation loss and DICE loss would be a much better choice, right? I am surprised your method is doing so well for the classification.

Yes, I believe removing the segmentation loss is a reasonable choice when there is no pixel-level ground truth in the training set. However, if such ground truth is available, I have not conducted ablation studies to verify this. It is possible that the segmentation process itself could contribute to improved classification performance.  Please feel free to contact me at your convenience if you have any further questions.

**Comment by YasuoFly** [comment]
@xiaozhen228 
Thank you very much!!! I feel clearer now.

**Comment by LonglongaaaGo** [comment]
Hey @xiaozhen228 cooooool!! I love that, thank you!!!
Thank you, @YasuoFly, as well.

**Comment by YasuoFly** [comment]
@xiaozhen228 

Sorry to bother you again. I've noticed that in the datasets.py file, when MyDataset reads data with mode set to 'train', it seems to be loading 'test' data instead. Is this an intended behavior or a bug?

<img width="600" height="481" alt="Image" src="https://github.com/user-attachments/assets/9da44f3c-5ede-4ffe-996e-e11abfd3a5ed" />

**Comment by xiaozhen228** [comment]
@YasuoFly  Hello, when mode = "train", setting meta_info = meta_info["test"] means selecting the test subset of the auxiliary dataset as auxiliary training data. This is because, in anomaly detection datasets, the train folder usually contains only normal samples, while the test folder includes both normal and real anomalous data, making it more suitable for use as an auxiliary training set. It should be noted that the auxiliary training and testing phases rely on different datasets.

**Comment by YasuoFly** [comment]
@xiaozhen228 Thank you for your quick reply.
I understand that the model is trained on dataset A and tested on dataset B, but in practice, the test set of A is used for training, while the training set of A is never used. Is my understanding correct?

**Comment by xiaozhen228** [comment]
> training set of A is never used

Yes, your understanding is correct. The training set of A contains only a large number of normal samples, which is often adopted in unsupervised anomaly detection settings.

**Comment by YasuoFly** [comment]
I get it, thanks a lot.

**Comment by xiaoming-678** [comment]
Hello, do you have any code for visualization? I'm a beginner, why do my visualization results look so strange?

<img width="518" height="518" alt="Image" src="https://github.com/user-attachments/assets/a8bffda6-2a1a-4253-a3bf-b0b3a33eaeb6" />这是我可视化的图
